Activity RefreshActitivityTree

AndreiTimisescu opened the pull request:
A client may change properties at runtime (while execution is paused), operation which requires a new call to CacheMetadata.
The Activity.RefreshActivityTree extension method provides this functionality.

https://uipath.atlassian.net/wiki/spaces/STUD/pages/87546724494/Hot+Reload+Investigation

gabriela-lungu-uip commented on src/UiPath.Workflow.Runtime/Tracking/TrackingProvider.cs:
```diff
@@ -324,6 +324,12 @@ private bool PostTrackingRecord(TrackingParticipant participant, RuntimeTracking
                 {
                     IAsyncResult result = participant.BeginTrack(preparedRecord, _timeoutHelper.RemainingTime(), PrepareAsyncCompletion(trackingCompleteCallback), this);
+
+                    if (preparedRecord is ActivityScheduledRecord record)
```
this affects existing workflows and functionalities

AndreiTimisescu replied:
It should do nothing as nothing changes on the regular path. But you are right, we may call this method only when needed.

AndreiTimisescu committed:
Old RuntimeArgument.BoundArgument.Expression.CacheId may cause validation errors.
Activity.RefreshActitivityTree refreshes the activity metadata
ActivityInfo.UpdateEnvironment updates the ActivityInfo runtime environment

gabriela-lungu-uip reviewed (approved):
UiPath.Workflow.Runtime code looks ok. This logic implemented in this PR is only called during Debug, so production/deployed workflows should not be impacted by the changes.
I did not review Debugger logic implemented in UiPath.Workflow

aoltean16 commented on src/UiPath.Workflow/Debugger/ActivityExtensions.cs:
```diff
@@ -0,0 +1,74 @@
+    /// ATTENTION - calling this method will generate a new activity.Id
+    /// </summary>
+    /// <returns>a list of validation errors</returns>
+    public static IList<ValidationError> RefreshActivityTree(this Activity activity)
```
Please rename this to something more meaningful, like RefreshActivityArguments. This way we "isolate" more what we're doing here, in case we change it in the future.

AndreiTimisescu committed:
review comments
